servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/ResolvingConnectionFactoryFilter.java

```diff
@@ -0,0 +1,139 @@
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import javax.annotation.Nullable;
+
+import static io.servicetalk.client.api.ServiceDiscovererEvent.Status.AVAILABLE;
+import static io.servicetalk.http.netty.GlobalDnsServiceDiscoverer.globalDnsServiceDiscoverer;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * A {@link ConnectionFactoryFilter} that will resolve the passed unresolved {@link InetSocketAddress} on each attempt
+ * to create a {@link ConnectionFactory#newConnection(Object, ContextMap, TransportObserver) new connection} using
+ * {@link GlobalDnsServiceDiscoverer#globalDnsServiceDiscoverer()}.
+ *
+ * @param <U> the type of address before resolution (unresolved address)
+ * @param <R> the type of address after resolution (resolved address)
+ */
+final class ResolvingConnectionFactoryFilter<U, R>
+        implements ConnectionFactoryFilter<R, FilterableStreamingHttpConnection> {
+
+    private final Function<R, U> toUnresolvedAddressMapper;
+    private final ServiceDiscoverer<U, R, ? extends ServiceDiscovererEvent<R>> serviceDiscoverer;
+
+    ResolvingConnectionFactoryFilter(
+            final Function<R, U> toUnresolvedAddressMapper,
+            final ServiceDiscoverer<U, R, ? extends ServiceDiscovererEvent<R>> serviceDiscoverer) {
+        this.toUnresolvedAddressMapper = requireNonNull(toUnresolvedAddressMapper);
+        this.serviceDiscoverer = requireNonNull(serviceDiscoverer);
+    }
+
+    @Override
+    public ConnectionFactory<R, FilterableStreamingHttpConnection> create(
+            final ConnectionFactory<R, FilterableStreamingHttpConnection> original) {
+        return new DelegatingConnectionFactory<R, FilterableStreamingHttpConnection>(original) {
+
+            @Override
+            @SuppressWarnings("unchecked")
+            public Single<FilterableStreamingHttpConnection> newConnection(final R address,
+                                           @Nullable final ContextMap context,
+                                           @Nullable final TransportObserver observer) {
+                final U unresolvedAddress = toUnresolvedAddressMapper.apply(address);
+                return serviceDiscoverer.discover(unresolvedAddress).takeAtMost(1).firstOrError()
+                        .flatMap(resolvedAddresses -> {
+                            @Nullable
+                            ServiceDiscovererEvent<R> resolved;
+                            if (resolvedAddresses.isEmpty()) {
+                                resolved = null;
+                            } else if (resolvedAddresses.size() == 1) {
+                                resolved = resolvedAddresses instanceof List ?
+                                        ((List<ServiceDiscovererEvent<R>>) resolvedAddresses).get(0) :
+                                        resolvedAddresses.stream().findFirst().orElse(null);
+                                if (!AVAILABLE.equals(resolved.status())) {
```

Comment:
if resolved is not a list and nothing found (so orElse(null)) triggers (I know it's very unlikely) would this resolved.status() lead to a NPE?

Reply:
This path executes only when `resolvedAddresses.size() == 1` but I will add a check anyway to avoid the worst case scenario if there is a tricky collection impl.
I used `resolvedAddresses.stream().findFirst().orElse(null)` as a short way instead of getting an Iterator and checking `hasNext()`.